### ISSUE TITLE
BUG: Two layers LSTM module couldn't be imported successfully.

### ISSUE THREAD

**Issue by FD-MINGJIE** [Unity-Technologies/barracuda-release#134]
@AlexRibard    
1. A **two layers** LSTM defined by nn.LSTM(num_layers = 2)  produced following bugs:
Asset import failed, "Assets/Resources/pose2image_lr_150e_9M_LSTM_football.onnx" > ArgumentException: An item with the same key has already been added. Key: recurrent_out_c
![image](https://user-images.githubusercontent.com/43561860/105030741-3327c680-5a8f-11eb-9fe7-226297ef0b29.png)
   I found one layer LSTM could be imported sucessfully, but two layers or stacted more one layers LSTM module did'n work!
2. The following image shows that **one layer** LSTM still has some bug like image shape of all zero.
![image](https://user-images.githubusercontent.com/43561860/105138140-72a5ef80-5b2f-11eb-8bb6-6d8036172858.png)
3. My LSTM input tensor's shape is (batch=1, 2, 105), but how could I **define three dimentional tensor** like 'new Tensor(1, 1, 2, 105, inputArray)'?
   My barracuda version is latest 1.2.1, and pytorch version is 1.8. Thanks for!
[two_layers_LSTM.zip](https://github.com/Unity-Technologies/barracuda-release/files/5840601/two_layers_LSTM.zip)

**Comment by amirebrahimi** [comment]
@FD-MINGJIE 

1. Yes, the one-layer LSTM is supported. Our LSTM support is not robust though, which explains why the two-layer doesn't import. I'll look into this further.
2. Re: the shape - you can interpret the 0s as *s meaning that the shape is resolved at runtime
3. You can create a `new Tensor(1, 1, 105, 2, inputArray)` - our layout is channels last, so NCH:1,2,105 -> N1WC:1,1,105,2

**Comment by FD-MINGJIE** [comment]
@amirebrahimi
     Thanks for your reply! 
     Now, the input of  one-layer LSTM can't match the tensor shape [1, 1 , 105, 2] in barracuda, which produced some errors in Dense layer：
![image](https://user-images.githubusercontent.com/43561860/105288473-d80ee480-5bf2-11eb-997e-d406cde419f1.png)
![image](https://user-images.githubusercontent.com/43561860/105293690-21abff00-5bf4-11eb-9b67-0accebf2afa6.png)

     My C# code in Unity are as following:
```
// choice 1
private float[] inputArray1D = new float[210]; // 105*2=210 
Tensor inputTensorFrames = new Tensor(1, 1, 105, 2, inputArray1D);  
// choice 2
private float[][] inputArray2D= new float[2][105]; 
Tensor inputTensorFrames = new Tensor(new TensorShape(1, 1, 105, 2 ), inputArray2D);
```
So, my problem is what shape should inputArray be ? Wish you give me an example for input of the one-layer LSTM.

I'm sure the main bug lies in the LSTM layer. If successfully imported, it'll display the shape of outputs rather than all zero.

So, I tested a model without LSTM using the same inputs. And it was not only imported successfully but also outputed expected results when inferring in barracuda like the following picture:
![image](https://user-images.githubusercontent.com/43561860/105326771-13b4a900-5c09-11eb-84e1-cba9e55c75f1.png)
```
// i take choice 2 as inputs 
private float[][] inputArray2D= new float[2][105]; 
Tensor inputTensorFrames = new Tensor(new TensorShape(1, 1, 105, 2 ), inputArray2D);
```
Ps:
1. The one-layer RNN also couldn't be imported.
2. I'd like to know when i can use dilated convolution (dilation > 1).

Thanks again!
Best wishes!
MJ
Annex：
(One-layer LSTM model)
[one_layer_LSTM.zip](https://github.com/Unity-Technologies/barracuda-release/files/5847294/one_layer_LSTM.zip)

**Comment by amirebrahimi** [comment]
@FD-MINGJIE - okay, so I've fixed this now and it will show up in the next preview release. In the meantime, if you want to make your own changes locally - this should work: 

https://github.com/Unity-Technologies/barracuda-release/blob/release/1.2.1/Barracuda/Runtime/ONNX/ONNXModelConverter.cs#L1203

`var baseLSTMOutputName = $"recurrent_out_{node.Name}";`

and

https://github.com/Unity-Technologies/barracuda-release/blob/release/1.2.1/Barracuda/Runtime/ONNX/ONNXModelConverter.cs#L1221

`var baseLSTMName = $"recurrent_in_{node.Name}";`

**Comment by amirebrahimi** [comment]
Re-opening because I missed your prior message, which I have reproduced and am currently looking into.

**Comment by FD-MINGJIE** [comment]
Thanks, the one-layer LSTM still has wrong inputs like above problem! Maybe you can try to run my model.

**Comment by FD-MINGJIE** [comment]
@amirebrahimi 
Is there any process about LSTM to share with me? I hope to adopt LSTM neural network to infer using barracuda engine as soon as possible.
Thanks!

**Comment by amirebrahimi** [comment]
Sorry, somehow I missed this post @FD-MINGJIE. This involved more refactors, but the good news is that we now have much better LSTM support. We have one minor release coming out soon and then this will land in the next release.

**Comment by amirebrahimi** [comment]
This work is now complete and will be included in v1.3.3.
